_posts/2016-04-30-akka-basics.md

```diff
@@ -0,0 +1,82 @@
+
+* Single responsibility principle. Failure management is delegated to supervisors, creating focused business actors. We agreed some time ago that lifecycle management, e.g. object creation, should be moved to factories, therefore an object is not responsible for creating itself. Resuming or restarting an object, after some failure happened, is part of that lifecycle management and Akka enforces you to move that responsibility to supervisors.
+
+* Default recovery strategies. It's important to understand which are our options after the system is under some failure condition. Is DB temporarily down? Has some input, in combination with the existing data, created an undesired state? Is that a bug? Do we need to abandon that particular request or that part of the system is messed up until some patch is applied? Those questions will determine our response to that incident, and Akka provides some [built-in strategies](http://doc.akka.io/docs/akka/2.4.4/scala/fault-tolerance.html#fault-tolerance-scala).
+
+```
+override val supervisorStrategy =
+  OneForOneStrategy(maxNrOfRetries = 10, withinTimeRange = 1 minute) {
+    case _: ArithmeticException      => Resume
+    case _: NullPointerException     => Restart
+    case _: IllegalArgumentException => Stop
+    case _: Exception                => Escalate
+  }
+```
+
+* Failure as first citizen. Some libraries or ecosystems hide failures through obscure APIs. Akka is usually deployed in distributed environments, therefore using unreliable resources like networks. That forces us to bring failure to the front page. Even in 'safer' environments as monoliths, failures are all around, so using toolkits like Akka is critical to create reliable and robust software.
+
+## Concurrency and parallelism
+
+With Akka we're not going to deal directly with threads, they're hidden under an abstraction layer. The backbone of an Akka app is the [Actor System](http://doc.akka.io/api/akka/2.0/akka/actor/ActorSystem.html):
+
+> An actor system is a hierarchical group of actors which share common configuration, e.g. dispatchers, deployments, remote capabilities and addresses. It is also the entry point for creating or looking up actors.
+
+A Dispatcher is also an Execution Context, so in the end is where the thread pool is located. Let's assume that your app uses a single dispatcher with 4 threads allocated. As you can see threads are a scarce resource, if we do blocking IO or heavy CPU work in some of the actors we'll use exclusively one of the threads, and you can see how quickly your service will run out of threads.
+
+The solution is wrapping those operation with constructs like Scala Futures and then provide a different execution context for those tasks.
+
+## Summary
+
+Akka is a fantastic toolkit full of well suited abstractions for modern problems. In next post we'll see with code how we can coordinate and supervise actors with the constraints of our example application.
```

Comment:
Maybe at the end of the series, I'll explain more Akka concepts in the next ones.